File: JAVA/Candy.java
public class Solution {
    public int candy(int[] ratings) {
        int len = ratings.length;
        if (len == 0) return 0;
        int num[] = new int[len];
        int cnt = 0;
        for (int i = 1; i < len; i++) {
            if (ratings[i] > ratings[i - 1]) {
                cnt++;
            }
            else {
                cnt = 0;
            }
            num[i] = cnt;
        }
        cnt = 0;
        for (int i = len - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                cnt++;
            }
            else {
                cnt = 0;
            }
            num[i] = Math.max(num[i], cnt);
        }
        int min = len;
        for (int i = 0; i < len; i++) {
            min += num[i];
        }
        return min;
    }
}

File: Candy.cpp
class Solution {
public:
    int candy(vector<int> &ratings) {
        int size = ratings.size();
        vector<int> candies(size);
        for (int i = 0; i < size; i++) {
            if (i == 0 || ratings[i] <= ratings[i-1]) {
                candies[i] = 1;
            } else {
                candies[i] = candies[i-1] + 1;
            }
        }
        int cur = 1;
        int result = 0;
        for (int i = size - 1; i >= 0; i--) {
            if (i == size - 1 || ratings[i] <= ratings[i + 1]) {
                cur = 1;
            } else {
                cur = cur + 1;
            }
            candies[i] = max(candies[i], cur);
            result += candies[i];
        }
        return result;
    }
};
